_posts/2020-09-10-centralised-log-management.markdown

```diff
@@ -0,0 +1,140 @@
+
+In the table we show the retention period - but how does data actually transition between the layers? There are many ways it can be achieved, but we will focus on the simplest.
+
+![Differing data lifecycles](/img/blog/clm/data-lifecycle.jpg)
+
+The beauty of this system is that we are not writing any code to move data around. This is important for a number of reasons
+
+* we gain resilience from not writing data / schema specific code
+* we leverage the cloud platform capabilities to guarantee "at least once" delivery semantics
+* we can set individual retention schedules dependent on the requirements of each data kind, but at the infrastructure / managed service layer, instead of our own code
+
+This makes for a sustainable growth trajectory for our log management system. As costs increase we can re-evaluate our needs and adjust retention schedules if necessary, or alternatively, bypass **hot storage** layers altogether.
+
+# Curation and Maintenance
+
+Now, you can point any of your scalable compute services mentioned above to the raw data on your warm storage bucket and search your data, but with any significant volume of data you will likely find that the performance is not what you would expect, and that's because the raw data that Firehose is writing to S3 is being written in micro batches (every 2 minutes or every 20Mb, as examples). 
+
+This is commonly referred to as the "small file problem" and though it's described as a problem it's more accurately a *Diseconomy of Scale* - as the number of files increases, the more *non value add* activities have to be run:
+
+* listing bucket files
+* bucket file fetching
+* content decryption
+* tcp negotiation
+* ssl handshaking
+
+For a few files the effect is negligible, however when you scale it to thousands of files and many concurrent users, it is very noticeable. The solution is to periodically create *rollups* <sup>[2]</sup> of those files. This is commonly done with Spark jobs that will aim to
+
+* rewrite data into columnar formats such as Parquet for optimized reading
+* compress the data for better performance and storage density
+* compact many small files in to larger files optimized for scalable compute query engines
+
+![Compacting small data files to larger aggregates](/img/blog/clm/data-compaction.png)
+
+There are many different ways to achieve this goal - a few are outlined below
+
+|method|pros|cons|
+|-|-|-|
+|EMR Cluster Spark Jobs| - pre-baked hadoop environment<br/>- transient clusters with spot instances could be cheaper|- high complexity|
+|Glue PySpark/Scala Jobs| - serverless<br/>- simple | - More expensive at scale|
+|Airflow / Luigi, etc.| - Full Workflow Management for your ETL | - not cloud native<br/>- operational complexity|
+|Kinesis Firehose Transformations| - Embedded transformation<br/>- Backup mode to allow capturing raw data|- Poor Documentation<br/>- May see poor economies of scale with high throughput|
+
```

Comment:
Does it need some kind of wrap / close out?

Reply:
yes